Humpty had a great fall

leorochael opened the pull request:
Drop humpty dependency.

Implements all wheel installation by subclassing `zc.buildout.easy_install.Instaler`.

Depends on merging:

- [ ] buildout/buildout#352

leorochael committed:
More surgical monkey-patch into setuptools

Patch only `distros_for_location` instead of both
`interpret_distro_name` and `EXTENSIONS`

leorochael committed:
Attempt to replace humpty

Things that work:

 * installation
 * entry points
 * scripts w/ or w/o entry points
 * namespace packages

Things that don't:

 * dependencies/requirements

leorochael commented:
@tseaver, how did you add "do not merge" to [a PR in the buildout repo](https://github.com/buildout/buildout/pull/344)?

mgedmin commented:
Someone with sufficient permissions has to [create the custom label](https://github.com/buildout/buildout.wheel/labels) first, and then someone with sufficient permissions can apply it to issues and pull requests.

mgedmin commented:
(Personally I like a `WIP:` prefix in the pull request title, to indicate that it's still work in progress, not ready to be merged.)

mgedmin commented on src/buildout/wheel/__init__.py:
```diff
@@ -1,56 +1,119 @@
+    We must self destruct to avoid having both the wheel and the generated egg
+    there at the same time."""
+    def __del__(self):
+        shutil.rmtree(self.location)
```
This delays the cleanup indefinitely if you use PyPy, and possibly never does it at all.

Is there no better place to do the cleanup deterministically?

leorochael replied:
> Is there no better place to do the cleanup deterministically?

There will be when we solve buildout/buildout#345.

mgedmin commented on src/buildout/wheel/__init__.py:
```diff
@@ -1,56 +1,119 @@
+        egg = self.distribution(location, metadata=metadata,
+                                class_=SelfDestructingEgg)
+        # Fix namespaces with missing __init__
+        if (sys.version_info < (3, 3) and
```
I was about to comment that "nobody supports Python 3.2 any more, why do we care", and then I realized this condition also matches Python 2.7.

leorochael commented:
So, I did reach a small snag.

`pkg_resources` is ready to read the `.dist-info` directory of a folder added to `sys.path`.

[Except if the folder on `sys.path` is called `something.egg`](https://github.com/pypa/setuptools/blob/92c5667e23e0c3a242182f3f869b86bad6d5a0ae/pkg_resources/__init__.py#L2008), in which case `pkg_resources` shortcuts and assume the metadata is in a directory called `EGG-INFO` inside it.

That would be find and dandy if all I had to do was rename  `somthing-version.dist-info` to `EGG-INFO', but if I do this, then I get a `Distribution` instead of a `DistInfoDistribution` for that package, and suddenly the metadata dependency can't be found anymore, as `pkg_resources` now expects it in `requirements.txt` instead of inside `METADATA` where wheels put it.

mgedmin commented:
> Except if the folder on sys.path is called something.egg, in which case pkg_resources shortcuts and assume the metadata is in a directory called EGG-INFO inside it.

Ah, good old https://github.com/pypa/setuptools/issues/462

leorochael commented:
There are a few possibilities from here:

- The Humpty way: reconstitute an actual `EGG-INFO` directory, rewriting `requirements.txt` and friends from the info inside `METADATA`.

- The Unpacked Wheel Way: Do not call the directory inside `eggs-directory` as `something.egg`, but add it to `sys.path` nonetheless.

 - We could call it `something.wheel`, like
   - `eggs/numpy-1.12.0-py2.7-linux-x86_64.wheel`
 - Or we could call it just `something`, like
   - `eggs/numpy-1.12.0-py2.7-linux-x86_64`

I'd rather do the Unpacked Wheel way, but it would require convincing `zc.buildout.easy_install` to take these non-egg directories into account.

@jimfulton, what do you think? Can you quickly point me to the part of `zc.buildout.easy_install` that takes the eggs inside `eggs/` into account while resolving requirements?

leorochael commented:
For reference, here is how [`Distribution` (`EGG-INFO`) parses dependencies](https://github.com/pypa/setuptools/blob/92c5667e23e0c3a242182f3f869b86bad6d5a0ae/pkg_resources/__init__.py#L2547).

And here is how [`DistInfoDistribution` does it](https://github.com/pypa/setuptools/blob/92c5667e23e0c3a242182f3f869b86bad6d5a0ae/pkg_resources/__init__.py#L2811).

jimfulton commented:
I think a requirement is that when setuptools looks at the result, it considers it a valid egg.  IOW, it isn't sufficient that it be importable. It must be recognized as a distribution later.

leorochael commented:
>  It must be recognized as a distribution later.

That's exactly the part that I want to hack.

`pkg_resources` can recognize all sorts of things as distributions. Reading the code I linked above I came under the impression that the file-name of the metadata directory (`something-version.dist-info`) may be enough to convince `pkg_resources` that they have a distribution, regardless of the extension of the directory containing it (as long as that directory is not called `something.egg`).

@jimfulton can you tell me how buildout passes the `buildout:eggs-directory` setting inside the `zc.buildout.easy_install` module? I'm having trouble locating it.

jimfulton commented:
On Mon, Feb 20, 2017 at 10:18 AM, Leonardo Rochael Almeida <
notifications@github.com> wrote:

> It must be recognized as a distribution later.
>
> That's exactly the part that I want to hack.
>
That word, "hack", in this context...



> pkg_resources can recognize all sorts of things as distributions. Reading
> the code I linked above I came under the impression that the file-name of
> the metadata directory (something-version.dist-info) may be enough to
> convince pkg_resources that they have a distribution, regardless of the
> extension of the directory containing it (as long as that directory is not
> called something.egg).
>
@jimfulton <https://github.com/jimfulton> can you tell me how buildout
> passes the buildout:eggs-directory setting inside the
> zc.buildout.easy_install module? I'm having trouble locating it.
>
It's passed as the ``dest`` argument to the ``install`` function.

leorochael committed:
Safer egg laying.

Avoid creating (and destroying) the egg in the download cache to
begin with.

Do it in a tempdir inside the download cache instead and destroy the
tmpdir afterwards.

leorochael committed:
Forget laying eggs from wheels: use the wheels

Unpack wheels directly as directories inside `eggs-directory`

Override `zc.buildout.easy_install.Installer` to feed the wheels inside
`eggs-directory` to `pkg_resources.Environment`.

That's enough for `pkg_resources` to read all the metadata.

Too bad... I liked the `lay_egg_into()` method name...

leorochael commented:
> > can you tell me how buildout passes the buildout:eggs-directory setting inside the zc.buildout.easy_install module? I'm having trouble locating it.
> 
> It's passed as the ``dest`` argument to the ``install`` function.

Thanks @jimfulton, the `Installer` class was the missing piece

> > > It must be recognized as a distribution later.
> >
> > That's exactly the part that I want to hack.
> That word, "hack", in this context...

Just a little hack... and it's working now:

This is a non-entry-point script:
```
$ cat bin/f2py 
#!/home/leonardo/Workspace/buildout.wheel/bin/python
# See http://cens.ioc.ee/projects/f2py2e/
from __future__ import division, print_function

import sys
sys.path[0:0] = [
  '/home/leonardo/Workspace/buildout.wheel/eggs/numpy-1.12.0-cp27-cp27mu-manylinux1_x86_64.whl',
  '/home/leonardo/Workspace/buildout.wheel/eggs/zest.releaser-6.9-py2.py3-none-any.whl',
  '/home/leonardo/Workspace/buildout.wheel/eggs/twine-1.8.1-py2.7.egg',
  '/home/leonardo/Workspace/buildout.wheel/lib/python2.7/site-packages',
  '/home/leonardo/Workspace/buildout.wheel/eggs/colorama-0.3.7-py2.7.egg',
  '/home/leonardo/Workspace/buildout.wheel/eggs/requests_toolbelt-0.7.1-py2.7.egg',
  '/home/leonardo/Workspace/buildout.wheel/eggs/requests-2.13.0-py2.7.egg',
  '/home/leonardo/Workspace/buildout.wheel/eggs/pkginfo-1.4.1-py2.7.egg',
  '/home/leonardo/Workspace/buildout.wheel/eggs/clint-0.5.1-py2.7.egg',
  '/home/leonardo/Workspace/buildout.wheel/eggs/args-0.1.0-py2.7.egg',
  ]
```

And this is an entry-point one:

```
$ cat bin/release 
#!/home/leonardo/Workspace/buildout.wheel/bin/python

import sys
sys.path[0:0] = [
  '/home/leonardo/Workspace/buildout.wheel/eggs/numpy-1.12.0-cp27-cp27mu-manylinux1_x86_64.whl',
  '/home/leonardo/Workspace/buildout.wheel/eggs/zest.releaser-6.9-py2.py3-none-any.whl',
  '/home/leonardo/Workspace/buildout.wheel/eggs/twine-1.8.1-py2.7.egg',
  '/home/leonardo/Workspace/buildout.wheel/lib/python2.7/site-packages',
  '/home/leonardo/Workspace/buildout.wheel/eggs/colorama-0.3.7-py2.7.egg',
  '/home/leonardo/Workspace/buildout.wheel/eggs/requests_toolbelt-0.7.1-py2.7.egg',
  '/home/leonardo/Workspace/buildout.wheel/eggs/requests-2.13.0-py2.7.egg',
  '/home/leonardo/Workspace/buildout.wheel/eggs/pkginfo-1.4.1-py2.7.egg',
  '/home/leonardo/Workspace/buildout.wheel/eggs/clint-0.5.1-py2.7.egg',
  '/home/leonardo/Workspace/buildout.wheel/eggs/args-0.1.0-py2.7.egg',
  ]

import zest.releaser.release

if __name__ == '__main__':
    sys.exit(zest.releaser.release.main())
```

Notice the `.whl` entries in the first two lines in each script. Those are actually directories. We can decide if we want a different extension so as not to confuse with zipped wheels.

The only difference between wheels and eggs is the message:

```
Installing 'numpy==1.12.0', 'zest.releaser'.
We have a develop egg: numpy 1.12.0
We have a develop egg: zest.releaser 6.9
```

Still need to test how the versions above interact with pinning, considering the `develop egg` message. They're not develop eggs at all, but...

Anyway, I'm happy... reasonably non-intrusive change, all things considered...

jimfulton commented:
>
> Notice the .whl entries in the first two lines in each script. Those are
> actually directories. We can decide if we want a different extension so as
> not to confuse with zipped wheels.
>
.ovo? :)


> The only difference between wheels and eggs is the message:
>
> Installing 'numpy==1.12.0', 'zest.releaser'.
> We have a develop egg: numpy 1.12.0
> We have a develop egg: zest.releaser 6.9
>
> Still need to test how the versions above interact with pinning,
> considering the develop egg message. They're not develop eggs at all,
> but...
>
Yeah. This worries me a little, just because we're a little "off the rails".

setuptools is deciding this is a develop dist for some reason and we're
treating it as such. We can check the extension and do something different,
although maybe we just want to not print the message.

> Anyway, I'm happy... reasonably non-intrusive change, all things
> considered...
>
Cool.

leorochael commented:
> > Notice the .whl entries in the first two lines in each script. Those are actually directories. We can decide if we want a different extension so as not to confuse with zipped wheels.
> 
> .ovo? :)

Wouldn't that be cute!

I settled on a `.dist` extension for now, but made nothing in the code treat it specially. It now scans all non-egg directories inside `eggs-directory` that have a `*.dist-info` inside and treat it same as eggs.

> Yeah. This worries me a little, just because we're a little "off the rails".

> setuptools is deciding this is a develop dist for some reason and we're treating it as such. We can check the extension and do something different, although maybe we just want to not print the message.

Treating it as a develop egg has all shorts of shortcuts and different branches inside `z.b.easy_install`, so I don't like it either.

This last commit adds an `pkg_resources.Environment` subclass to go along with the `z.b.easy_install.Installer` override that knows how to scan dists inside `eggs-directory` and reports the dists it find int there as `EGG_DIST`.

At this point I'm happy with the state of this PR and would invide further reviews so that we could merge it.

But other than that, I have a hunch I'm investigating that might allow us to get rid of the `wheel_to_egg` hook and concentrate all monkey-patches/subclassing on setuptools itself.

leorochael committed:
Replace `.whl` with `.dist` for wheel directories

They're not just unzipped wheels. File have been moved around, so
we should not reuse the extension.

Instead, let's call them just `dist`.

Also, feed everything in `eggs-directory`  that can be a dist to
`Environment`, both .egg and .dist entries.

Finally, move the tempdir cleanup handling to `wheel_to_egg` and the
extension unloading to get rid of `SelfDestructingDistro`.

leorochael committed:
Look, Ma! no wheel_to_egg!

Handle the wheel conversion inside Installer._call_easy_install

This means we don't need to delay tempdirs cleanup.

jimfulton commented on src/buildout/wheel/__init__.py:
```diff
@@ -1,56 +1,222 @@
+
+
+def _get_dest_dist_paths(dest):
+    p = os.path
```
Ugh.  I hate renames.  Why not just:

    from os.path import join

and use join below?

leorochael replied:
> Ugh. I hate renames. Why not just:
> 
>      from os.path import join
> 
> and use join below?

sure, it's just that I'm using many functions from `os.path` in this block and, in general, I don't like importing many naked names (`join`, `dirname`) at the top of the module.

jimfulton commented on src/buildout/wheel/__init__.py:
```diff
@@ -1,56 +1,222 @@
+    return pkg_resources._by_version_descending(set(eggs + dists))
+
+
+class Environment(pkg_resources.Environment):
```
I'm worried about these cross-org subclasses.

At least setuptools is pretty stable and unlikely to change, but it should be a priority to get rid of these and get them into setuptools, however, using wheels as an installation format isn't exactly a setuptools goal.

leorochael replied:
> I'm worried about these cross-org subclasses.

There are two parts to the `Environment` subclass.

The first is that, as of today, Buildout passes the `eggs-directory` to `Environment` as a single path, instead of passing all eggs as separate paths.

I think this is abusing the `Environment` a little, since `eggs-directory` is never going to be directly in `sys.path`, but `Environment` treats it as it it would.

If we change buildout to feed all eggs individually to `Environment`, the `scan()` override wouldn't be necessary.

The second aspect is a bit trickier. Buildout treats specially all dists that have `EGG_DIST` precedence. It assumes they're things it can chose to use or ignore and fetch something more appropriate. Whereas any other `precedence` value is treated by Buildout as things that it can't change (i.e. they're either develop eggs, or come from the environment).

In order not to override the `Environment.add` method, the `Installer` would need to check each `dist.location` against the `eggs-directory` instead of checking the precedence.

That or we would need to go through all dists in `Environment` and mark those inside `eggs-directory` as `EGG_DIST` after each `scan`.

So, if we're willing to change buildout a bit, the `Environment` subclass would be unnecessary.

jimfulton commented on src/buildout/wheel/__init__.py:
```diff
@@ -1,56 +1,222 @@
+            shutil.rmtree(tmp)
+
+
+class WheelDir(wheel.install.WheelFile):
```
This subclass is even more concerning because wheel seems like it's still evolving.

Can't this be done with composition instead?

leorochael replied:
Sure, the only overridden method here was in `__init__`, and that was just to change the default value of a parameter.

jimfulton reviewed (commented):
I don't like the coupling to the pypa packages.  This feels very likely to change and break us.

It's *almost* (whimper) hard to believe that it's so hard to generate egg meta data that this is easier than a generating eggs.

I appreciate the hard work you've done, but this leaves me queasy.

I wonder what other folks think.

leorochael reviewed (commented):
> I don't like the coupling to the pypa packages. This feels very likely to change and break us.

None of the couplings are particularly hard, except to setuptools, to which Buildout is already married for better or worse.

> It's almost (whimper) hard to believe that it's so hard to generate egg meta data that this is easier than a generating eggs.

Generating metadata in the EGG-INFO format and naming our installed packages as `.egg` is not particularly hard, `humpty` does it... I just felt that, with setuptools being able to understand `wheel` metadata as is (as long it's not in a directory called `*.egg`), we shouldn't *havet to* re-generate it.

> I appreciate the hard work you've done, but this leaves me queasy.

> At least setuptools is pretty stable and unlikely to change, but it should be a priority to get rid of these and get them into setuptools, however, using wheels as an installation format isn't exactly a setuptools goal.

The coupling, and the style of this PR was done in a way to point the direction of an eventual support for wheels in setuptools itself. I can rework it to make it less queasy (less subclassing, pushing some of the code back to buildout itself).

As for setuptools using wheels as an install format, reading the issues in the project I don't see any resistance to the idea, they're just waiting for someone to cough up the PR.

leorochael committed:
Use a WheelFile adapter instead of subclass

Using composition, instead of inheritance, replace the WheelFile subclass
with a WheelFile adapter.

Helps reducing the code of the `distros_for_location()` monkeypatch

leorochael committed:
Look, Ma! No foreign subclasses!

Stop subclassing `pkg_resources.Environment`.

Just override `Installer._call_easy_install()` to unpack wheels
inside `eggs-directory` such that `Installer` knows to find them.

leorochael commented:
@jimfulton, By using the encapsulating methods added in buildout/buildout#325, we can get rid of all subclasses except the Installer.

On the other hand, we still don't need to do anything in `wheel_to_egg()`, so maybe we can consider dropping that hook altogether.

jimfulton commented on .travis.yml:
```diff
@@ -6,7 +6,7 @@ python:
   - 3.4
 install:
-  - pip install -U pip setuptools zc.buildout
+  - pip install -U pip setuptools https://github.com/buildout/buildout/archive/encapsulate-easy-install-env.zip#egg=zc.buildout
```
Wow, I hadn't seen this trick before.  Clever.

jimfulton commented on src/buildout/wheel/__init__.py:
```diff
@@ -1,56 +1,230 @@
+
+    def _env_get_dists_for(self, project_name):
+        dists = orig_Installer._env_get_dists_for(self, project_name)
+        return [self._maybe_eggify(dist) for dist in dists]
```
As mentioned in the review of the corresponding buildout branch, I'd like to go ahead and move these over there.

leorochael replied:
Will do.

jimfulton commented on src/buildout/wheel/__init__.py:
```diff
@@ -1,56 +1,230 @@
-    ):
-    if '.whl' in location:
-        path = parse.urlparse(location).path
+class WheelInstaller(object):
```
I'm actually reviewing this class. :)  I should and maybe will later.

leorochael replied:
There's really nothing magical about it now. To understand what it does I recommend just searching for it's instantiations and following its uses.

jimfulton replied:
I wasn't suggesting there was anything magical about it.  I was merely asserting my laziness and my confidence in you. :)

jimfulton reviewed (commented):
Thanks!

leorochael commented on src/buildout/wheel/__init__.py:
```diff
@@ -1,56 +1,230 @@
+       `DistInfoDistribution`.
+    """
+
+    _dist_extension = '.dist'
```
I'm considering renaming this to `.pydist`, `.dist` is too generic and meaningless in a context where buildout is automating all sorts of things and not just Python.

Specially if Buildout starts going in a less-Python/setuptools-specific direction and we end up renaming the `eggs` to something else.

jimfulton replied:
I still like .ovo, but I defer.

I don't have any intention of running away from eggs (now that we have wheel support).  While I hope we end up with a virtualenv/pip recipe, I think eggs are a useful alternate model.

(I also hope that we find ways to improve tool support for eggs, to respond to your excellent points on the pip-recipe issue.)

leorochael commented:
Hi @jimfulton. This extension is a lot thinner now.

leorochael committed:
TMP: tweak travis to use the zc.buildout branch we depend on.

Please drop this commit before merging.

leorochael commented:
> I still like .ovo, but I defer.

Oh... I really thought you were just joking... didn't get you were only half joking.

The extension has no bearing on the functionality, so if you want `.ovo` I can make it so. It's literally a one-word-change.

In any case, I think this PR is good for merging, pending the merging of  buildout/buildout#352, and whatever changes are still done to it.

I specially like it how I could throw a bunch of code away now that `_call_easy_install()` does nothing but land unpacked distributions in requested directory and return their paths.

> I don't have any intention of running away from eggs (now that we have wheel support). While I hope we end up with a virtualenv/pip recipe, I think eggs are a useful alternate model.

The changes in buildout/buildout#352 show that the egg model can be usefully applied to wheels or any other distribution format, as long as there is a way to unpack it into directory that can be added to `sys.path` while preserving the metadata in a format readable by `pkg_resources`.

jimfulton commented:
On Wed, Mar 1, 2017 at 12:58 PM, Leonardo Rochael Almeida <
notifications@github.com> wrote:

> I still like .ovo, but I defer.
>
> Oh... I really thought you were just joking... didn't get you were only
> half joking.
>
I was completely serious.  These "are" eggs, IMO, to the degree that eggs
are the install analog of jar files.

> The extension has no bearing on the functionality, so if you want .ovo I
> can make it so. It's literally a one-word-change.
>
+1

> In any case, I think this PR is good for merging, pending the merging of
> buildout/buildout#352 <https://github.com/buildout/buildout/pull/352>,
> and whatever changes are still done to it.
>
I'm essentially reviewing these together. I'll take another look at both
soon.



> I specially like it how I could throw a bunch of code away now that
> _call_easy_install() does nothing but land unpacked distributions in
> requested directory and return their paths.
>

Cool.


> I don't have any intention of running away from eggs (now that we have
> wheel support). While I hope we end up with a virtualenv/pip recipe, I
> think eggs are a useful alternate model.
>
> The changes in buildout/buildout#352
> <https://github.com/buildout/buildout/pull/352> show that the egg model
> can be usefully applied to wheels or any other distribution format, as long
> as there is a way to unpack it into directory that can be added to
> sys.path while preserving the metadata in a format readable by
> pkg_resources.
>
Yup.

Thanks for your work!

leorochael commented:
@jimfulton, I squeezed one more refactoring out of `._get_dist()`.

Now the hook can be installed not by monkey-patching, but by adding data to a global dict.

The `wheel_to_egg` function continues existing for the benefit of early adopters, and works as before, except for the warning it emits when used.

jimfulton commented:
LGTM, other than the failing tests. :)

Thanks!

leorochael committed:
Look, Ma! No Buildout subclasses either!

After further refactoring of the Buildout API itself, now we can just
add a, say, wheel unpacker to the "registry" of distribution
unpacking functions, instead of subclassing `easy_install.Installer`.

leorochael committed:
Use `.ovo` as the non-egg dist extension.

jimfulton commented:
Holding off on the merge until we release version of buildout that this depends on.

leorochael commented:
> Holding off on the merge until we release version of buildout that this depends on.

Makes sense. We should include in this PR a `<=` requirement to the released version of Buildout that supports it before merging.

grzn commented:
now that buildout-2.9.2 is released, when are you planning to merge and release this PR?

Commit:
Use released zc.buildout

jimfulton commented:
Thanks for the poke. I just updated the travis config to use the released zc.buildout.  When the tests are green, I'll merge and make a release.

@leorochael could you send me your pypi id?

leorochael commented:
@jimfulton: leorochael

jimfulton commented:
@leorochael I was about to release, but realized there wasn't a change log entry.

Could you add one please (on master).  You're a pypi owner now, so you can add a release, but I'm happy to as well, once the change log is updated.

leorochael commented:
@jimfulton,

I've added `CHANGES.rst` entries, but didn't release. I'm in a bit of a hurry now, and I'm afraid I'll make a mistake if I don't pay enough attention while releasing.

If I get some more calm time and you haven't released yet, I'll try my hand, but fell free to cut the release.

jimfulton commented:
I'll make a release tomorrow. I've finally started using zest released,
which makes it pretty automatic.

Jim

On Mar 11, 2017 6:21 PM, "Leonardo Rochael Almeida" <
notifications@github.com> wrote:

> @jimfulton <https://github.com/jimfulton>,
>
> I've added CHANGES.rst entries, but didn't release. I'm in a bit of a
> hurry now, and I'm afraid I'll make a mistake if I don't pay enough
> attention while releasing.
>
> If I get some more calm time and you haven't released yet, I'll try my
> hand, but fell free to cut the release.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/buildout/buildout.wheel/pull/9#issuecomment-285908246>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAvKd6dolzizP9WaBb_tzrDtT3_YWw-0ks5rkyxcgaJpZM4MF4Ux>
> .
>

jimfulton commented:
On Sat, Mar 11, 2017 at 9:45 PM, Jim Fulton <jim@jimfulton.info> wrote:

> I'll make a release tomorrow. I've finally started using zest released,
> which makes it pretty automatic.
>

Done, 0.2.0.

Thanks

Jim


>
> Jim
>
> On Mar 11, 2017 6:21 PM, "Leonardo Rochael Almeida" <
> notifications@github.com> wrote:
>
>> @jimfulton <https://github.com/jimfulton>,
>>
>> I've added CHANGES.rst entries, but didn't release. I'm in a bit of a
>> hurry now, and I'm afraid I'll make a mistake if I don't pay enough
>> attention while releasing.
>>
>> If I get some more calm time and you haven't released yet, I'll try my
>> hand, but fell free to cut the release.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/buildout/buildout.wheel/pull/9#issuecomment-285908246>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAvKd6dolzizP9WaBb_tzrDtT3_YWw-0ks5rkyxcgaJpZM4MF4Ux>
>> .
>>
>


-- 
Jim Fulton
http://jimfulton.info
